### ISSUE TITLE
The MP JWT 1.2 TCK EmptyTokenTest fails with Error 500

### ISSUE THREAD

**Issue by clarkek123** [OpenLiberty/open-liberty#14011]
**Describe the bug**
The MP JWT 1.2 TCK EmptyTokenTest fails with error 500 when either 401 or RC 200 is expected. 

If there is a stack trace, please include the FULL stack trace (without any `[internal classes]` lines in it). To find the full stack trace, you may need to check in `$WLP_OUTPUT_DIR/messages.log`

**Steps to Reproduce**
1. Run MP JWT 1.2 EmptyTokenTest as part of aud_env runner

**Expected behavior**
The tests run expect either RC 200 or RC 401 but are getting error 500

invalidToken - expected RC 401 and got 500
validToken - expected RC 200 and got 500
emptyToken - expected RC 200 and got 500

The additional message logged is as follows (and I expect that the test will need to tolerate this message as expected.   Please verify:
CWWKS5604E: A JsonWebToken Principal can't be injected because one is not available. Protect the requesting resource so authentication occurs before the resource is accessed

**Diagnostic information:**
 - OpenLiberty Version: development master SHA 8e5e85a76c10ce159aec58ac84f1d5f62fa21503
 - Java Version: Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.17.0, JRE 11 Mac OS X amd64-64-Bit Compressed References 20191016_371 (JIT enabled, AOT enabled)
OpenJ9 - 77c1cf708
OMR - 20db4fbc
JCL - 2a7af5674b based on jdk-11.0.5+10)
 - server.xml configuration (WITHOUT sensitive information like passwords)

```XML
<server description="mp_jwt">

   <include location="../fatTestPorts.xml"/>  
   <!--  
     <httpEndpoint id="defaultHttpEndpoint"
                  httpPort="8010"
                  httpsPort="8020" />
   -->
   
    <!-- This prevents apps from granting themselves this permission -->
    <javaPermission className="java.security.AllPermission"  name="*" actions="*" />
    
    <featureManager>        
        <feature>componenttest-1.0</feature>
        <feature>jaxrs-2.1</feature>        
        <feature>mpJwt-1.2</feature>
         <feature>appSecurity-2.0</feature>        
        <feature>localConnector-1.0</feature>
        <feature>arquillian-support-1.0</feature>
    </featureManager>

    <mpJwt  groupNameAttribute="groups" id="myMpJwt" audiences="s6BhdRkqt3, s6BhdRkqt3.2" /> 
  
    <!-- <logging maxFileSize="20" maxFiles="10" traceFileName="trace.log" traceSpecification="com.ibm.ws.jaxrs*=all:com.ibm.ws.webcontainer.security.*=all:com.ibm.ws.security.*=all:com.ibm.wsspi.webcontainer*=all"/>  --> 
      <logging maxFileSize="20" maxFiles="10" traceFileName="trace.log" traceSpecification="*=info:com.ibm.websphere.security.*=all:com.ibm.ws.cdi*=all:com.ibm.ws.jaxrs*=all:com.ibm.ws.security.*=all:com.ibm.ws.webcontainer.security.*=all:com.ibm.ws.weld*=all:com.ibm.wsspi.webcontainer*=all:io.openliberty.security.*=all" /> 
 
    <keyStore id="defaultKeyStore" password="xxxxx/>
	
   
     <webApplication id="RolesAllowedTest" location="${server.config.dir}/apps/RolesAllowedTest.war" name="RolesAllowedTest">
       <application-bnd>
           <security-role name="Token2Role">
              <group access-id="group:https://server.example.com/group1.2" name="group1.2"/>  
               <group access-id="group:https://server.example.com/group1.2" name="Token2Role"/>   <!--  new  -->     
           </security-role>
           <security-role name="Group1MappedRole">
              <group access-id="group:https://server.example.com/group1" name="group1"/>   
              <group access-id="group:https://server.example.com/group1" name="group1MappedRole"/>       <!--  new  -->    
           </security-role>
           <security-role name="Tester">              
               <group access-id="group:https://server.example.com/Tester" name="Tester"/>  
            </security-role>
            <security-role name="Echoer">
                <group access-id="group:https://server.example.com/Echoer" name="Echoer"/> 
            </security-role>
       </application-bnd> 
    </webApplication> 
    

</server>
```

**Additional context**
Messages.log contains:
```
[9/17/20, 15:36:06:770 CDT] 00000035 com.ibm.ws.app.manager.AppMessageHelper                      A CWWKZ0001I: Application EmptyTokenTest started in 4.635 seconds.
[9/17/20, 15:36:08:779 CDT] 0000002f .ws.org.apache.cxf.cxf.core.3.2:1.0.45.202009161052(id=119)] I Setting the server's publish address to be /
[9/17/20, 15:36:08:992 CDT] 0000002f com.ibm.ws.webcontainer.servlet                              I SRVE0242I: [EmptyTokenTest] [/EmptyTokenTest] [org.eclipse.microprofile.jwt.tck.container.jaxrs.TCKApplication]: Initialization successful.
[9/17/20, 15:36:09:258 CDT] 0000002f com.ibm.ws.security.mp.jwt.principal.PrincipalBean           E CWWKS5604E: A JsonWebToken Principal can't be injected because one is not available. Protect the requesting resource so authentication occurs before the resource is accessed.
[9/17/20, 15:36:09:281 CDT] 0000002f com.ibm.ws.security.mp.jwt.principal.PrincipalBean           E CWWKS5604E: A JsonWebToken Principal can't be injected because one is not available. Protect the requesting resource so authentication occurs before the resource is accessed.
[9/17/20, 15:36:09:300 CDT] 0000002f org.apache.cxf.jaxrs.utils.JAXRSUtils                        E No message body writer has been found for class org.glassfish.json.JsonObjectBuilderImpl$JsonObjectImpl, ContentType: text/plain
[9/17/20, 15:36:09:537 CDT] 00000035 com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI            E CWWKS5523E: The MicroProfile JWT feature cannot authenticate the request because the token that is included in the request cannot be validated. JWT processing failed. Additional details: [[17] Unable to process JOSE object (cause: org.jose4j.lang.JoseException: Invalid JOSE Compact Serialization. Expecting either 3 or 5 parts for JWS or JWE respectively but was 1.): something]
[9/17/20, 15:36:09:577 CDT] 00000035 com.ibm.ws.security.mp.jwt.principal.PrincipalBean           E CWWKS5604E: A JsonWebToken Principal can't be injected because one is not available. Protect the requesting resource so authentication occurs before the resource is accessed.
[9/17/20, 15:36:09:584 CDT] 00000035 org.apache.cxf.jaxrs.utils.JAXRSUtils                        E No message body writer has been found for class org.glassfish.json.JsonObjectBuilderImpl$JsonObjectImpl, ContentType: text/plain
[9/17/20, 15:36:10:569 CDT] 0000003b org.apache.cxf.jaxrs.utils.JAXRSUtils                        E No message body writer has been found for class org.glassfish.json.JsonObjectBuilderImpl$JsonObjectImpl, ContentType: text/plain
[9/17/20, 15:36:11:144 CDT] 00000031 com.ibm.ws.config.xml.internal.ConfigRefresher               A CWWKG0016I: Starting server configuration update.
[9/17/20, 15:36:11:146 CDT] 00000031 com.ibm.ws.config.xml.internal.XMLConfigParser               A CWWKG0028A: Processing included configuration resource: /Users/kristenclarke/libertyGit/open-liberty/dev/build.image/wlp/usr/servers/fatTestPorts.xml
[9/17/20, 15:36:11:147 CDT] 00000031 com.ibm.ws.config.xml.internal.XMLConfigParser               A CWWKG0028A: Processing included configuration resource: /Users/kristenclarke/libertyGit/open-liberty/dev/build.image/wlp/usr/servers/fatTestCommon.xml
[9/17/20, 15:36:11:165 CDT] 00000031 com.ibm.ws.config.xml.internal.ConfigRefresher               A CWWKG0017I: The server configuration was successfully updated in 0.022 seconds.
[9/17/20, 15:36:11:219 CDT] 00000030 com.ibm.ws.http.internal.VirtualHostImpl                     A CWWKT0017I: Web application removed (default_host): http://192.168.1.19:8010/EmptyTokenTest/
[9/17/20, 15:36:11:225 CDT] 00000030 com.ibm.ws.webcontainer.servlet                              I SRVE0253I: [EmptyTokenTest] [/EmptyTokenTest] [org.eclipse.microprofile.jwt.tck.container.jaxrs.TCKApplication]: Destroy successful.
[9/17/20, 15:36:11:295 CDT] 00000030 com.ibm.ws.app.manager.AppMessageHelper                      A CWWKZ0009I: The application EmptyTokenTest has stopped successfully.
```

Test details:
The setup for the test builds the following to package with EmptyTokenTest.war

```Java
    @Deployment
    public static WebArchive createDeployment() {
        URL config = InvalidTokenTest.class.getResource("/META-INF/microprofile-config-publickey-location.properties");
        URL publicKey = InvalidTokenTest.class.getResource("/publicKey.pem");
        return ShrinkWrap
            .create(WebArchive.class, "EmptyTokenTest.war")
            .addAsManifestResource(new StringAsset(MpJwtTestVersion.MPJWT_V_1_2.name()), MpJwtTestVersion.MANIFEST_NAME)
            .addAsResource(publicKey, "/publicKey.pem")
            .addClass(TCKApplication.class)
            .addClass(EmptyTokenEndpoint.class)
            .addAsWebInfResource("beans.xml", "beans.xml")
            .addAsManifestResource(config, "microprofile-config.properties");
    }
```

The tests run expect either RC 200 or RC 401 but are getting error 500

invalidToken - expected RC 401 and got 500
validToken - expected RC 200 and got 500
emptyToken - expected RC 200 and got 500

The additional message logged is as follows (and I expect that the test will need to tolerate this message as expected.   Please verify:
CWWKS5604E: A JsonWebToken Principal can't be injected because one is not available. Protect the requesting resource so authentication occurs before the resource is accessed

```Java
   @RunAsClient
    @Test(groups = TEST_GROUP_JAXRS,
          description = "Validate that an empty JWT in injected in the endpoint")
    public void emptyToken() {
        String uri = baseURL.toExternalForm() + "endp/verifyEmptyToken";
        WebTarget echoEndpointTarget = ClientBuilder.newClient().target(uri);
        Response response = echoEndpointTarget.request(TEXT_PLAIN).get();
        Assert.assertEquals(response.getStatus(), HttpURLConnection.HTTP_OK);
        String replyString = response.readEntity(String.class);
        JsonReader jsonReader = Json.createReader(new StringReader(replyString));
        JsonObject reply = jsonReader.readObject();
        Assert.assertTrue(reply.getBoolean("pass"));
    }

    @RunAsClient
    @Test(groups = TEST_GROUP_JAXRS,
          description = "Validate that a token sent to an unauthenticated / unauthorized endpoint is verified")
    public void invalidToken() {
        String uri = baseURL.toExternalForm() + "endp/verifyEmptyToken";
        WebTarget echoEndpointTarget = ClientBuilder.newClient()
                                                    .target(uri)
                                                    .queryParam("input", "hello");
        Response response = echoEndpointTarget
            .request(TEXT_PLAIN)
            .header(HttpHeaders.AUTHORIZATION, "Bearer " + "something")
            .get();
        Assert.assertEquals(response.getStatus(), HttpURLConnection.HTTP_UNAUTHORIZED);
    }

    @RunAsClient
    @Test(groups = TEST_GROUP_JAXRS,
          description = "Validate that a token sent to an unauthenticated / unauthorized endpoint is verified and " +
                        "injected as non-empty")
    public void validToken() throws Exception {
        String token = TokenUtils.generateTokenString("/Token1.json");
        String uri = baseURL.toExternalForm() + "endp/verifyNonEmptyToken";
        WebTarget echoEndpointTarget = ClientBuilder.newClient()
                                                    .target(uri)
                                                    .queryParam("input", "hello");
        Response response = echoEndpointTarget
            .request(TEXT_PLAIN)
            .header(HttpHeaders.AUTHORIZATION, "Bearer " + token)
            .get();
        Assert.assertEquals(response.getStatus(), HttpURLConnection.HTTP_OK);
        String replyString = response.readEntity(String.class);
        JsonReader jsonReader = Json.createReader(new StringReader(replyString));
        JsonObject reply = jsonReader.readObject();
        Assert.assertTrue(reply.getBoolean("pass"));
    }
}
```

**Comment by ayoho** [comment]
Need to investigate whether a behavior change from 1.1 is expected, per a new section added to https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/interoperability.asciidoc#injection-of-jsonwebtoken:
> If there is no JWT in the request, an empty JsonWebToken is injected, which means all method calls to this token return null. Note that MP JWT will still perform Authentication and Authorization if the endpoint requires these verifications. Effectively, the injected empty token is only visible on unauthenticated and unauthorized endpoints.
>
> If a JWT is sent to an endpoint that does not require Authentication and/or Authorization then it still must be verified before it can be accessed via JsonWebToken interface.

In MP JWT 1.1, it sounds like we were simply skipping the injection of a `JsonWebToken` object for unprotected resources. Now we might need to actually inject an empty object in those cases.

**Comment by ayoho** [comment]
Our runtime actually does appear to be working correctly for this test. There actually appears to be a test bug where the JAX-RS client used by this TCK test class is set to accept `text/plain` in the HTTP response, however the corresponding endpoint is sending back JSON. Changing the tests to accept the correct media type appears to make the test successful.

I opened https://github.com/eclipse/microprofile-jwt-auth/pull/212 to make the change to the TCK, and this issue should be fixed once that PR is merged and a new release candidate is available for us to use.

**Comment by ayoho** [comment]
https://github.com/eclipse/microprofile-jwt-auth/pull/212 was merged into that repo's master branch on Oct 12 at 7:56 PM CT.

**Comment by clarkek123** [comment]
This is fixed with the MP JWT TCK1.2 - RC2 which has been uploaded to artifactory and new tests are running successfully. So I'm closing this issue.
